protocol/tx/nonce.go

```diff
@@ -8,16 +8,23 @@ type nonce struct {
 		TimeRange bc.Hash
 		ExtHash   bc.Hash
 	}
+
+	// TimeRange contains (a pointer to) the manifested entry
+	// corresponding to body.TimeRange.
+	TimeRange *timeRange
 }
 
 func (nonce) Type() string         { return "nonce1" }
 func (n *nonce) Body() interface{} { return n.body }
 
 func (nonce) Ordinal() int { return -1 }
 
-func newNonce(p program, tr bc.Hash) *nonce {
+func newNonce(p program, tr *timeRange) *nonce {
 	n := new(nonce)
 	n.body.Program = p
-	n.body.TimeRange = tr
+	if tr != nil {
```

Comment:
Hmmmm what do you think about adding a case to entryID to return a zero hash when the entry is nil? Then we might be able to get rid of most of these if statements in the constructors.